### PR TITLE
Refactor ShaderProgram creation

### PR DESCRIPTION
**Pull request by lilleyse**
For #640

### CODE REVIEW

File: Source/Renderer/ShaderProgram.js [hunk 1]
```diff
@@ -56,6 +58,30 @@ define([
...
+    ShaderProgram.fromCache = function(options) {
+        options = defaultValue(options, defaultValue.EMPTY_OBJECT);
+
+        //>>includeStart('debug', pragmas.debug);
+        if (!defined(options.context)) {
+            throw new DeveloperError('options.context is required.');
+        }
+        //>>includeEnd('debug');
+
+        return options.context._shaderCache.getShaderProgram(options);
```
Comment (pjcozzi) [inline Source/Renderer/ShaderProgram.js]: `Context` has a read-only `shaderCache` property.  Let's use it.

File: Source/Renderer/ShaderProgram.js [hunk 2]
```diff
@@ -56,6 +58,30 @@ define([
...
+    ShaderProgram.replaceCache = function(options) {
+        options = defaultValue(options, defaultValue.EMPTY_OBJECT);
+
+        //>>includeStart('debug', pragmas.debug);
+        if (!defined(options.context)) {
+            throw new DeveloperError('options.context is required.');
+        }
+        //>>includeEnd('debug');
+
+        return options.context._shaderCache.replaceShaderProgram(options);
```
Comment (pjcozzi) [inline Source/Renderer/ShaderProgram.js]: Here too.

File: Specs/Renderer/FramebufferSpec.js [hunk 3]
```diff
@@ -11,6 +11,7 @@ defineSuite([
         'Renderer/PixelDatatype',
         'Renderer/Renderbuffer',
         'Renderer/RenderbufferFormat',
+         'Renderer/ShaderProgram',
```
Comment (pjcozzi) [inline Specs/Renderer/FramebufferSpec.js]: Remove the extra leading whitespace.

**Comment by pjcozzi** [comment]
Looks good, just those trivial comments.

**Comment by lilleyse** [comment]
Updated.

**Comment by pjcozzi** [comment]
Very nice.

**Comment by mramato** [comment]
@pjcozzi While these recent cleanup pull requests all look good, have we profiled at all to see if it has increased garbage collection?  On the surface, it looks like we are creating a lot more temporary objects than we used to (as options constructors).  I've had a long standing to-do item to either have less internal options objects (when all parameters are required anyway) or at least make these calls use a module-level scratch variable to avoid repeated allocations (but obviously we should profile some different use cases first to see if it really is even a problem).

**Comment by pjcozzi** [comment]
I knew someone would ask, but I am not remotely concerned about the extra temporary objects - these are rare and heavy operations, e.g., they happen when preparing a new terrain tile, not when issuing a draw call.

**Pull request merged**